pyproject.toml

```diff
@@ -7,3 +7,57 @@ disable_error_code = ["union-attr", "annotation-unchecked", "import-untyped"]
 explicit_package_bases = true
 warn_unused_ignores = true
 exclude = ['tests']
+
+[tool.poetry]
+name = "tagstudio"
+version = "9.3.2"
+packages = [
+  {include = "tagstudio/tag_studio.py"},
+  {include = "tagstudio/src/core"},
+  {include = "tagstudio/src/qt"},
+  {include = "tagstudio/resources"},
+  {include = "tagstudio/__init__.py"}
+]
+
+description = "User-Focused Photo & File Management System"
+readme = "README.md"
+authors = ["Travis Abendshien"]
+license = "GPL-3.0-only"
+repository = "https://github.com/TagStudioDev/TagStudio"
+maintainers = [] # Maybe someone can find a way to export this from GitHub?
+
+[tool.poetry.urls]
+Discord = "https://discord.gg/hRNnVKhF2G"
+
+[tool.poetry.dependencies]
+python = ">=3.11 <3.13"
+humanfriendly = "10.0"
+opencv_python=">=4.8.0.74,<=4.9.0.80"
+Pillow="10.3.0"
```

Comment:
This is a comment for the rest of the pinned versions and this does not reflect a change I want to enforce, but an honest comment. 

`pyproject.toml` in a library should focus on getting the newest version of a package that it can. This means that you should not pin versions here. 

`requirements.txt` is the place to pin a version, as this file that the application should use. this is the file dependabot will use to check for security issues and in another scenario, this is the file that docker would use as well. 

Here lies one of the issues with poetry, it is not 100% compatible with the python ecosystem, and you cant use a constraint file to enforce application versions during install. It will follow the lock file. 

Micromanaging the pyproject toml files pinned versions is a little hassle, but it can ofcourse be done, this is a cost someone has to take; I'm in favor if you can. I use poetry at work so that all developers can have the same experience, at the cost of some extra work for one. 

TL;DR
I don't like pinned versions in pyproject.toml, sometimes that is the only way until poetry complies with pip.

Reply:
> `pyproject.toml` in a library should focus on getting the newest version of a package that it can. This means that you should not pin versions here.

I think it's exact the opposite. Defining dependencies too loosely seems to lead to unforeseen problems and non-reproducible builds, especially in Python and Javascript projects. Ultimately, that is what the lock file is trying to avoid. If necessary, it can be updated quickly at any time.

> `requirements.txt` is the place to pin a version, as this file that the application should use. this is the file dependabot will use to check for security issues and in another scenario, this is the file that docker would use as well.

Just an idea, but maybe it's possible to use a git hook to generate a `requirements.txt` from the `pyproject.toml`.
(or can an automatic build system generate a temporary one and call dependabot?)

Reply:
I disagree that pyproject.toml should focus on getting the newest version of a package. The important thing is that the project works, not that the latest version is downloaded. From time to time a pr can bump versions after testing. Also in your other comment you say you actually don't like specifying `^` versions. 

Why would you pin versions in requirements.txt and not in pyproject.toml? That just does not make any sense.  The versions should be the same and that is why I explained in the CONTRIBUTING.md  how to generate requirements.txt from pyproject.toml.

If it was possible I would not even have a requirements.txt but we have to have it of course.

Reply:
I think having a git hook to update requirement.txt from the pyproject.toml is a good idea. I also thought of that but I have never set up git hooks before.

Reply:
The Poetry project seems to provide some git hooks.

E.g. I have seen a [poetry-export git hook](https://python-poetry.org/docs/pre-commit-hooks/#poetry-export).

But it requires to install a poetry plugin.
Not sure if it is suitable at all or even working on NixOS systems.

I don't really have any experience with this, there may be better tools out there for creating _requirements.txt_ files.

Maybe a temporary _requirements.txt_ during the build would be sufficient, which would then be used for testing while in CI/CD.

Or perhaps there are better alternatives to `dependabot` that would work without any problems with just a `pyproject.toml`, the industry seems to shift towards it anyway. Unless the _requirements.txt_ is absolutely needed for something.

Reply:
Cool. But this could also be another pr?